rectools/model_selection/cross_validate.py

```diff
@@ -0,0 +1,139 @@
+    user_features_new = item_features_new = None
+    if user_features is not None:
+        user_features_new = user_features.take(user_id_map.get_external_sorted_by_internal())  # 2x internal
+    if item_features is not None:
+        item_features_new = item_features.take(item_id_map.get_external_sorted_by_internal())  # 2x internal
+    dataset = Dataset(
+        user_id_map=user_id_map,
+        item_id_map=item_id_map,
+        interactions=interactions_train,
+        user_features=user_features_new,
+        item_features=item_features_new,
+    )
+    return dataset
+
+
+def cross_validate(  # pylint: disable=too-many-locals
+    dataset: Dataset,
+    splitter: Splitter,
+    metrics: tp.Dict[str, MetricAtK],
+    models: tp.Dict[str, ModelBase],
+    k: int,
+    filter_viewed: bool,
+    items_to_recommend: tp.Optional[ExternalIds] = None,
+) -> tp.Dict[str, tp.Any]:
+    """
+    Run cross validation on multiple models with multiple metrics.
+
+    Parameters
+    ----------
+    dataset : Dataset
+        Dataset with full data.
+    splitter : Splitter
+        Instance of any `rectools.model_selection.Splitter` subclasses.
+    metrics : dict(str -> MetricAtK)
+        Dict of initialized metric objects to calculate,
+        where key is metric name and value is metric object.
+    models : dict(str -> ModelBase)
+        Dict of initialized model objects to fit and measure quality,
+        where key is metric name and value is metric object.
+    k : int
+        Derived number of recommendations for every user.
+        For some models actual number of recommendations may be less than `k`.
+    filter_viewed : bool
+        Whether to filter from recommendations items that user has already interacted with.
+    items_to_recommend : array-like, optional, default None
+        Whitelist of external item.
+        If given, only these items will be used for recommendations.
+
+    Returns
+    -------
+    dict
+        Dictionary with structure
+        {
+            "splits": [
+                {"i_split": 1, <split_info>},
+                {"i_split": 2, <split_info>},
+                ...
+            ],
+            "metrics": [
+                {"model": "model_1", "i_split": 1, <metrics>},
+                {"model": "model_1", "i_split": 2, <metrics>},
+                {"model": "model_2", "i_split": 1, <metrics>},
+                ...
+            ]
+        }
+    """
+    interactions = dataset.interactions
+
+    split_iterator = splitter.split(interactions, collect_fold_stats=True)
+
+    split_infos = []
+    metrics_all = []
+
+    for train_ids, test_ids, split_info in split_iterator:
+        split_infos.append(split_info)
+
+        interactions_df_train = interactions.df.iloc[train_ids]  # 1x internal
+        fold_dataset = gen_2x_internal_ids_dataset(interactions_df_train, dataset.user_features, dataset.item_features)
```

Comment:
Let's write a comment here why we need to do this: we need to avoid fitting models on sparse matrices with all zero rows/columns => we need to create a fold dataset which contains only hot users and items for current training.
Should we rename the func to something more meaningful? Like "get_filtered_train_dataset". The main goal is to filter, mapping is just unavoidable for this. And possibly comment that mapping is done to new x2 internal ids